bundles/org.eclipse.osgi/container/src/org/eclipse/osgi/internal/serviceregistry/ServiceUse.java

```diff
@@ -28,22 +32,66 @@
  * @ThreadSafe
  */
 public class ServiceUse<S> {
+	/** BundleContext associated with this service use */
+	final BundleContextImpl context;
 	/** ServiceDescription of the registered service */
 	final ServiceRegistrationImpl<S> registration;
+	final Debug debug;
 
 	/** bundle's use count for this service */
 	/* @GuardedBy("this") */
 	private int useCount;
 
+	/** lock for this service */
+	private ReentrantLockExt lock = new ReentrantLockExt();
+	public static final int DEADLOCK = 1001;
+
+	private int lockTimeout = 10;
+
 	/**
 	 * Constructs a service use encapsulating the service object.
 	 *
 	 * @param   context bundle getting the service
 	 * @param   registration ServiceRegistration of the service
 	 */
 	ServiceUse(BundleContextImpl context, ServiceRegistrationImpl<S> registration) {
-		this.useCount = 0;
+		this.context = context;
 		this.registration = registration;
+		this.useCount = 0;
+		this.debug = context.getContainer().getConfiguration().getDebug();
+	}
+
+	void runWithLock(Runnable runnable) {
+		callWithLock(Executors.callable(runnable));
+	}
+
+	<T> T callWithLock(Callable<T> callable) {
+		try {
+			if (lock.tryLock(getLockTimeout(), TimeUnit.SECONDS)) {
+				return callable.call();
+			}
+			throw new ServiceException(
+					"Failed to acquire lock within try period, Lock, id:" + System.identityHashCode(lock) + ", " //$NON-NLS-1$ //$NON-NLS-2$
+							+ lock.toString(),
+					DEADLOCK); // $NON-NLS-1$
+		} catch (InterruptedException e) {
+			Thread.currentThread().interrupt(); // To appease sonar
+			return null;
+		} catch (Exception e) {
+			throw new RuntimeException("Callable threw exception", e); //$NON-NLS-1$
+		} finally {
+			if (lock.isHeldByCurrentThread()) {
+				lock.unlock();
+			}
+		}
+	}
+
+	int getLockTimeout() {
+		return lockTimeout;
+	}
+
+	ReentrantLockExt getLock() {
```

Comment:
```suggestion
	ReentrantLock getLock() {
```

There is no need to expose the impl details of this lock (that it is some specific subclass.)